File: wearable/wear-os/managers/SensorManager.kt
/**
 * AquaMetric WearOS - Sensor Manager
 * 
 * Android SensorManager を使用した100Hzセンサーデータ収集
 * SwimBIT仕様: 100Hz サンプリングレート必須
 */

package com.aquametric.wearable.managers

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentLinkedQueue

// ========================================
// Constants
// ========================================

object SensorConstants {
    /** SwimBIT仕様の100Hz */
    const val SAMPLING_RATE_HZ = 100
    
    /** マイクロ秒単位のサンプリング間隔 */
    const val SAMPLING_PERIOD_US = 1_000_000 / SAMPLING_RATE_HZ
    
    /** バッファサイズ（5秒分） */
    const val BUFFER_SIZE = 500
    
    /** 1サンプルあたりのバイト数 */
    const val BYTES_PER_SAMPLE = 32
}

// ========================================
// Data Models
// ========================================

/**
 * 1サンプルのセンサーデータ
 */
data class SensorSample(
    val timestamp: Long,        // ナノ秒
    val accelerometerX: Float,
    val accelerometerY: Float,
    val accelerometerZ: Float,
    val gyroscopeX: Float,
    val gyroscopeY: Float,
    val gyroscopeZ: Float
) {
    /**
     * バイナリ形式に変換（32バイト）
     */
    fun toBytes(): ByteArray {
        val buffer = ByteBuffer.allocate(SensorConstants.BYTES_PER_SAMPLE)
            .order(ByteOrder.LITTLE_ENDIAN)
        
        buffer.putLong(timestamp)
        buffer.putFloat(accelerometerX)
        buffer.putFloat(accelerometerY)
        buffer.putFloat(accelerometerZ)
        buffer.putFloat(gyroscopeX)
        buffer.putFloat(gyroscopeY)
        buffer.putFloat(gyroscopeZ)
        // 4バイトのパディング
        buffer.putInt(0)
        
        return buffer.array()
    }
}

/**
 * セッション情報
 */
data class SwimSession(
    val sessionId: String,
    val startTime: Long,
    var endTime: Long? = null,
    val poolLengthM: Int = 25,
    var sampleCount: Int = 0
)

// ========================================
// Ring Buffer
// ========================================

/**
 * スレッドセーフなリングバッファ
 */
class RingBuffer<T>(private val capacity: Int) {
    private val buffer = ConcurrentLinkedQueue<T>()
    
    val count: Int get() = buffer.size
    
    fun write(element: T) {
        if (buffer.size >= capacity) {
            buffer.poll()
        }
        buffer.offer(element)
    }
    
    fun readAll(): List<T> {
        val result = buffer.toList()
        buffer.clear()
        return result
    }
    
    fun clear() {
        buffer.clear()
    }
}

// ========================================
// Sensor Manager
// ========================================

/**
 * センサーデータ収集マネージャー
 */
class AquaMetricSensorManager(private val context: Context) : SensorEventListener {
    
    private val sensorManager: SensorManager = 
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    
    private val accelerometer: Sensor? = 
        sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    
    private val gyroscope: Sensor? = 
        sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
    
    private val buffer = RingBuffer<SensorSample>(SensorConstants.BUFFER_SIZE)
    
    private var isRecording = false
    private var sessionStartTime: Long = 0
    
    // 一時的なセンサー値の保持
    private var lastAccel = floatArrayOf(0f, 0f, 0f)
    private var lastGyro = floatArrayOf(0f, 0f, 0f)
    private var lastAccelTime: Long = 0
    private var lastGyroTime: Long = 0
    
    // バッチ通知用のFlow
    private val _batchFlow = MutableSharedFlow<List<SensorSample>>()
    val batchFlow: SharedFlow<List<SensorSample>> = _batchFlow
    
    /**
     * センサー記録を開始
     */
    fun startRecording(): Boolean {
        if (accelerometer == null || gyroscope == null) {
            android.util.Log.e("SensorManager", "Required sensors not available")
            return false
        }
        
        if (isRecording) return true
        
        sessionStartTime = System.nanoTime()
        
        // 100Hz でセンサー登録
        sensorManager.registerListener(
            this,
            accelerometer,
            SensorConstants.SAMPLING_PERIOD_US
        )
        sensorManager.registerListener(
            this,
            gyroscope,
            SensorConstants.SAMPLING_PERIOD_US
        )
        
        isRecording = true
        android.util.Log.i("SensorManager", "Started recording at 100Hz")
        return true
    }
    
    /**
     * センサー記録を停止
     */
    fun stopRecording(): List<SensorSample> {
        if (!isRecording) return emptyList()
        
        sensorManager.unregisterListener(this)
        isRecording = false
        
        val remainingSamples = buffer.readAll()
        buffer.clear()
        
        android.util.Log.i("SensorManager", "Stopped recording. Samples: ${remainingSamples.size}")
        return remainingSamples
    }
    
    // ========================================
    // SensorEventListener
    // ========================================
    
    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                lastAccel = event.values.clone()
                lastAccelTime = event.timestamp
            }
            Sensor.TYPE_GYROSCOPE -> {
                lastGyro = event.values.clone()
                lastGyroTime = event.timestamp
            }
        }
        
        // 両方のセンサーデータが揃ったらサンプルを作成
        if (lastAccelTime > 0 && lastGyroTime > 0 &&
            kotlin.math.abs(lastAccelTime - lastGyroTime) < 5_000_000) { // 5ms以内
            
            val sample = SensorSample(
                timestamp = System.nanoTime(),
                accelerometerX = lastAccel[0],
                accelerometerY = lastAccel[1],
                accelerometerZ = lastAccel[2],
                gyroscopeX = lastGyro[0],
                gyroscopeY = lastGyro[1],
                gyroscopeZ = lastGyro[2]
            )
            
            buffer.write(sample)
            
            // バッファが満杯になったら通知
            if (buffer.count >= SensorConstants.BUFFER_SIZE) {
                val batch = buffer.readAll()
                // Coroutine scope で emit
                kotlinx.coroutines.runBlocking {
                    _batchFlow.emit(batch)
                }
            }
        }
    }
    
    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // 精度変更は記録のみ
        android.util.Log.d("SensorManager", "Accuracy changed: $sensor -> $accuracy")
    }
}

// ========================================
// Storage Manager
// ========================================

/**
 * ローカルストレージへのセンサーデータ保存
 */
class StorageManager(private val context: Context) {
    
    private val sessionsDir: File by lazy {
        File(context.filesDir, "sessions").also { it.mkdirs() }
    }
    
    /**
     * セッションデータをファイルに保存
     */
    fun saveSession(sessionId: String, samples: List<SensorSample>): File {
        val file = File(sessionsDir, "$sessionId.bin")
        
        file.outputStream().buffered().use { output ->
            samples.forEach { sample ->
                output.write(sample.toBytes())
            }
        }
        
        android.util.Log.i("StorageManager", "Saved ${samples.size} samples to ${file.name}")
        return file
    }
    
    /**
     * 未同期セッションの一覧を取得
     */
    fun getPendingSessions(): List<File> {
        return sessionsDir.listFiles()
            ?.filter { it.extension == "bin" }
            ?: emptyList()
    }
    
    /**
     * セッションファイルを削除（同期完了後）
     */
    fun deleteSession(sessionId: String): Boolean {
        val file = File(sessionsDir, "$sessionId.bin")
        return file.delete()
    }
}

// ========================================
// Session Manager
// ========================================

/**
 * 水泳セッションの管理
 */
class SessionManager(context: Context) {
    
    private val sensorManager = AquaMetricSensorManager(context)
    private val storageManager = StorageManager(context)
    
    private var currentSession: SwimSession? = null
    private val currentSessionSamples = mutableListOf<SensorSample>()
    
    val isSessionActive: Boolean get() = currentSession != null
    
    /**
     * セッション開始
     */
    fun startSession(poolLengthM: Int = 25): String {
        val sessionId = UUID.randomUUID().toString()
        currentSession = SwimSession(
            sessionId = sessionId,
            startTime = System.currentTimeMillis(),
            poolLengthM = poolLengthM
        )
        currentSessionSamples.clear()
        
        val started = sensorManager.startRecording()
        android.util.Log.i("SessionManager", "Session $sessionId started: $started")
        
        return sessionId
    }
    
    /**
     * セッション終了
     */
    fun endSession(): String? {
        val session = currentSession ?: return null
        
        // 残りのサンプルを取得
        val remainingSamples = sensorManager.stopRecording()
        currentSessionSamples.addAll(remainingSamples)
        
        // 終了時刻を記録
        session.endTime = System.currentTimeMillis()
        session.sampleCount = currentSessionSamples.size
        
        // ファイルに保存
        storageManager.saveSession(session.sessionId, currentSessionSamples)
        
        val endedSessionId = session.sessionId
        currentSession = null
        currentSessionSamples.clear()
        
        return endedSessionId
    }
    
    /**
     * 経過時間（秒）を取得
     */
    val elapsedTimeSeconds: Long
        get() {
            val session = currentSession ?: return 0
            return (System.currentTimeMillis() - session.startTime) / 1000
        }
    
    /**
     * 現在のサンプル数を取得
     */
    val sampleCount: Int
        get() = currentSessionSamples.size
}

File: wearable/apple-watch/Managers/SensorManager.swift
/// AquaMetric Apple Watch - Sensor Manager
///
/// CoreMotion を使用した100Hzセンサーデータ収集
/// 
/// SwimBIT仕様: 100Hz サンプリングレート必須

import Foundation
import CoreMotion

// MARK: - Constants

enum SensorConstants {
    /// SwimBIT仕様の100Hz
    static let samplingRate: Double = 100.0
    static let updateInterval: TimeInterval = 1.0 / samplingRate
    
    /// バッファサイズ（5秒分）
    static let bufferSize: Int = 500
    
    /// 32 bytes per sample (timestamp + 6 floats + padding)
    static let bytesPerSample: Int = 32
}

// MARK: - Sensor Data Model

/// 1サンプルのセンサーデータ
struct SensorSample {
    let timestamp: UInt64        // ナノ秒
    let accelerometerX: Float
    let accelerometerY: Float
    let accelerometerZ: Float
    let gyroscopeX: Float
    let gyroscopeY: Float
    let gyroscopeZ: Float
    
    /// バイナリ形式に変換（32バイト）
    func toBytes() -> Data {
        var data = Data(capacity: SensorConstants.bytesPerSample)
        
        var ts = timestamp
        data.append(Data(bytes: &ts, count: 8))
        
        var ax = accelerometerX, ay = accelerometerY, az = accelerometerZ
        data.append(Data(bytes: &ax, count: 4))
        data.append(Data(bytes: &ay, count: 4))
        data.append(Data(bytes: &az, count: 4))
        
        var gx = gyroscopeX, gy = gyroscopeY, gz = gyroscopeZ
        data.append(Data(bytes: &gx, count: 4))
        data.append(Data(bytes: &gy, count: 4))
        data.append(Data(bytes: &gz, count: 4))
        
        // Padding to 32 bytes
        data.append(contentsOf: [UInt8](repeating: 0, count: 4))
        
        return data
    }
}

// MARK: - Ring Buffer

/// リングバッファ - メモリ効率的なセンサーデータ蓄積
class RingBuffer<T> {
    private var buffer: [T?]
    private var writeIndex: Int = 0
    private var readIndex: Int = 0
    private(set) var count: Int = 0
    let capacity: Int
    
    init(capacity: Int) {
        self.capacity = capacity
        self.buffer = [T?](repeating: nil, count: capacity)
    }
    
    func write(_ element: T) {
        buffer[writeIndex] = element
        writeIndex = (writeIndex + 1) % capacity
        count = min(count + 1, capacity)
    }
    
    func readAll() -> [T] {
        var result: [T] = []
        for i in 0..<count {
            let index = (readIndex + i) % capacity
            if let element = buffer[index] {
                result.append(element)
            }
        }
        return result
    }
    
    func clear() {
        writeIndex = 0
        readIndex = 0
        count = 0
        buffer = [T?](repeating: nil, count: capacity)
    }
}

// MARK: - Sensor Manager Protocol

protocol SensorManagerDelegate: AnyObject {
    func sensorManager(_ manager: SensorManager, didCollectBatch samples: [SensorSample])
    func sensorManager(_ manager: SensorManager, didEncounterError error: Error)
}

// MARK: - Sensor Manager

/// センサーデータ収集マネージャー
class SensorManager {
    
    // MARK: Properties
    
    private let motionManager = CMMotionManager()
    private let queue = OperationQueue()
    private var buffer: RingBuffer<SensorSample>
    private var isRecording = false
    private var sessionStartTime: UInt64 = 0
    
    weak var delegate: SensorManagerDelegate?
    
    // MARK: Initialization
    
    init() {
        buffer = RingBuffer(capacity: SensorConstants.bufferSize)
        queue.name = "com.aquametric.sensor"
        queue.maxConcurrentOperationCount = 1
    }
    
    // MARK: Recording Control
    
    /// センサー記録を開始
    func startRecording() -> Bool {
        guard motionManager.isDeviceMotionAvailable else {
            print("Device motion not available")
            return false
        }
        
        guard !isRecording else {
            return true
        }
        
        // 100Hz設定
        motionManager.deviceMotionUpdateInterval = SensorConstants.updateInterval
        
        // 開始時刻を記録
        sessionStartTime = currentNanoseconds()
        
        motionManager.startDeviceMotionUpdates(to: queue) { [weak self] motion, error in
            guard let self = self, let motion = motion else {
                if let error = error {
                    self?.delegate?.sensorManager(self!, didEncounterError: error)
                }
                return
            }
            
            self.processSensorData(motion)
        }
        
        isRecording = true
        print("Started recording at 100Hz")
        return true
    }
    
    /// センサー記録を停止
    func stopRecording() -> [SensorSample] {
        guard isRecording else { return [] }
        
        motionManager.stopDeviceMotionUpdates()
        isRecording = false
        
        // 残りのデータを返す
        let remainingSamples = buffer.readAll()
        buffer.clear()
        
        print("Stopped recording. Total samples in buffer: \(remainingSamples.count)")
        return remainingSamples
    }
    
    // MARK: Private Methods
    
    private func processSensorData(_ motion: CMDeviceMotion) {
        let sample = SensorSample(
            timestamp: currentNanoseconds(),
            accelerometerX: Float(motion.userAcceleration.x),
            accelerometerY: Float(motion.userAcceleration.y),
            accelerometerZ: Float(motion.userAcceleration.z),
            gyroscopeX: Float(motion.rotationRate.x),
            gyroscopeY: Float(motion.rotationRate.y),
            gyroscopeZ: Float(motion.rotationRate.z)
        )
        
        buffer.write(sample)
        
        // バッファが満杯になったらフラッシュ
        if buffer.count >= SensorConstants.bufferSize {
            let batch = buffer.readAll()
            buffer.clear()
            delegate?.sensorManager(self, didCollectBatch: batch)
        }
    }
    
    private func currentNanoseconds() -> UInt64 {
        return UInt64(Date().timeIntervalSince1970 * 1_000_000_000)
    }
}

// MARK: - Storage Manager

/// ローカルストレージへのセンサーデータ保存
class StorageManager {
    
    private let fileManager = FileManager.default
    
    var sessionsDirectory: URL {
        let docs = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        return docs.appendingPathComponent("sessions", isDirectory: true)
    }
    
    init() {
        // セッションディレクトリを作成
        try? fileManager.createDirectory(at: sessionsDirectory, withIntermediateDirectories: true)
    }
    
    /// セッションデータをファイルに保存
    func saveSession(sessionId: String, samples: [SensorSample]) throws -> URL {
        let fileURL = sessionsDirectory.appendingPathComponent("\(sessionId).bin")
        
        var data = Data()
        for sample in samples {
            data.append(sample.toBytes())
        }
        
        try data.write(to: fileURL)
        print("Saved \(samples.count) samples to \(fileURL.lastPathComponent)")
        
        return fileURL
    }
    
    /// 未同期セッションの一覧を取得
    func getPendingSessions() -> [URL] {
        guard let files = try? fileManager.contentsOfDirectory(
            at: sessionsDirectory,
            includingPropertiesForKeys: nil
        ) else {
            return []
        }
        
        return files.filter { $0.pathExtension == "bin" }
    }
    
    /// セッションファイルを削除（同期完了後）
    func deleteSession(at url: URL) throws {
        try fileManager.removeItem(at: url)
    }
}

// MARK: - Session Manager

/// 水泳セッションの管理
class SessionManager: SensorManagerDelegate {
    
    private let sensorManager = SensorManager()
    private let storageManager = StorageManager()
    
    private var currentSessionId: String?
    private var currentSessionSamples: [SensorSample] = []
    private var sessionStartTime: Date?
    
    var isSessionActive: Bool { currentSessionId != nil }
    
    init() {
        sensorManager.delegate = self
    }
    
    // MARK: Session Control
    
    /// セッション開始
    func startSession(poolLengthM: Int = 25) -> String {
        let sessionId = UUID().uuidString
        currentSessionId = sessionId
        currentSessionSamples = []
        sessionStartTime = Date()
        
        let started = sensorManager.startRecording()
        print("Session \(sessionId) started: \(started)")
        
        return sessionId
    }
    
    /// セッション終了
    func endSession() -> String? {
        guard let sessionId = currentSessionId else { return nil }
        
        // 残りのサンプルを取得
        let remainingSamples = sensorManager.stopRecording()
        currentSessionSamples.append(contentsOf: remainingSamples)
        
        // ファイルに保存
        do {
            _ = try storageManager.saveSession(
                sessionId: sessionId,
                samples: currentSessionSamples
            )
        } catch {
            print("Failed to save session: \(error)")
        }
        
        let endedSessionId = currentSessionId
        currentSessionId = nil
        currentSessionSamples = []
        sessionStartTime = nil
        
        return endedSessionId
    }
    
    /// 経過時間を取得
    var elapsedTime: TimeInterval {
        guard let start = sessionStartTime else { return 0 }
        return Date().timeIntervalSince(start)
    }
    
    /// 現在のサンプル数を取得
    var sampleCount: Int {
        return currentSessionSamples.count
    }
    
    // MARK: SensorManagerDelegate
    
    func sensorManager(_ manager: SensorManager, didCollectBatch samples: [SensorSample]) {
        currentSessionSamples.append(contentsOf: samples)
        
        // ログ出力（デバッグ用）
        print("Collected batch: \(samples.count) samples. Total: \(currentSessionSamples.count)")
    }
    
    func sensorManager(_ manager: SensorManager, didEncounterError error: Error) {
        print("Sensor error: \(error)")
    }
}
